merge Txpool into s3

rlan35 opened the pull request:
Cherry pick txpool logic from master to s3

Tested locally:
====== WALLET BALANCES ======
Using local profile for wallet
Account: one1spshr72utf6rwxseaz339j09ed8p6f8ke370zj:
    Balance in Shard 0:  183.38074238, nonce: 10
    Balance in Shard 1:  0.0000, nonce: 0
Using local profile for wallet
Account: one1uyshu2jgv8w465yc8kkny36thlt2wvel89tcmg:
    Balance in Shard 0:  0.0000, nonce: 0
    Balance in Shard 1:  184.952170952, nonce: 10
Using local profile for wallet
Account: one1r4zyyjqrulf935a479sgqlpa78kz7zlcg2jfen:
    Balance in Shard 0:  1.0000, nonce: 0
    Balance in Shard 1:  1.0000, nonce: 0
====== RESULTS ======
PASS

LeoHChen commented on node/worker/worker.go:
```diff
@@ -420,8 +384,8 @@ func New(config *params.ChainConfig, chain *core.BlockChain, engine consensus_en
 	}
-	worker.gasFloor = 500000000000000000
-	worker.gasCeil = 1000000000000000000
+	worker.gasFloor = 80000000
```
any comments on why change this number?

rlan35 replied:
Previous setting doesn't make sense when we first launch mainnet. It was too high. The 80m to 120m limit is 10x that of Ethereum. So we have 10x more capacity than ethereum. And it translate to roughly max 5700 normal txns in a block

LeoHChen commented on node/worker/worker.go:
```diff
@@ -420,8 +384,8 @@ func New(config *params.ChainConfig, chain *core.BlockChain, engine consensus_en
-	worker.gasFloor = 500000000000000000
-	worker.gasCeil = 1000000000000000000
+	worker.gasFloor = 80000000
+	worker.gasCeil = 120000000
```
any reason on using this number?

LeoHChen commented on core/blockchain.go:
```diff
@@ -1136,6 +1136,7 @@ func (bc *BlockChain) InsertChain(chain types.Blocks) (int, error) {
 	n, events, logs, err := bc.insertChain(chain)
 	bc.PostChainEvents(events, logs)
 	if err == nil {
+		// TODO: incorporate these into insertChain
```
any impact? can you open a ticket on this?

rlan35 replied:
the TODO is already done in master. But the code at cherrypick doesn't

LeoHChen commented on core/tx_pool.go:
```diff
@@ -301,12 +301,11 @@ func (pool *TxPool) loop() {
-				//pool.reset(head.Header(), ev.Block.Header())
-				//head = ev.Block
-
+				if pool.chainconfig.IsS3(ev.Block.Epoch()) {
```
which signer shall we use?

rlan35 replied:
S3 is the condition we used to include homestead when we did the upgrade before.

LeoHChen commented on go.mod:
```diff
@@ -62,6 +62,7 @@ require (
 	golang.org/x/lint v0.0.0-20190409202823-959b441ac422
 	golang.org/x/sys v0.0.0-20190801041406-cbf593c0f2f3 // indirect
 	golang.org/x/tools v0.0.0-20190924052046-3ac2a5bbd98a
+	google.golang.org/appengine v1.4.0 // indirect
```
this is not needed.

rlan35 replied:
Reverted.

LeoHChen commented on node/node.go:
```diff
@@ -286,23 +284,19 @@ func (node *Node) tryBroadcast(tx *types.Transaction) {
 // Add new staking transactions to the pending staking transaction list.
 func (node *Node) addPendingStakingTransactions(newStakingTxs staking.StakingTransactions) {
-	txPoolLimit := core.ShardingSchedule.MaxTxPoolSizeLimit()
+	txPoolLimit := 1000 // TODO: incorporate staking txn into TxPool
```
use a constant please.

LeoHChen commented on node/node.go:
```diff
@@ -313,7 +307,7 @@ func (node *Node) addPendingStakingTransactions(newStakingTxs staking.StakingTra
 	}
 	node.pendingStakingTxMutex.Unlock()
-	utils.Logger().Info().Int("length of newStakingTxs", len(newStakingTxs)).Int("totalPending", len(node.pendingTransactions)).Msg("Got more staking transactions")
+	utils.Logger().Info().Int("length of newStakingTxs", len(newStakingTxs)).Int("totalPending", len(node.pendingStakingTransactions)).Msg("Got more staking transactions")
```
do we have staking transaction?

rlan35 replied:
Staking is not supported in s3. But somehow the initial staking txns skeleton code is in s3. It's no-op so shouldn't be a problem.

LeoHChen commented on node/node_handler_test.go:
```diff
@@ -34,11 +36,10 @@ func TestAddNewBlock(t *testing.T) {
-		func([]types.RPCTransactionError) {},
-	)
+	txs := make(map[common.Address]types.Transactions)
+	stks := types2.StakingTransactions{}
```
we don't have staking tx in s3, right?

LeoHChen commented on node/node_handler_test.go:
```diff
@@ -69,12 +70,10 @@ func TestVerifyNewBlock(t *testing.T) {
-	)
-
+	txs := make(map[common.Address]types.Transactions)
+	stks := types2.StakingTransactions{}
```
we don't have staking tx in s3, right?

LeoHChen commented on node/node_newblock.go:
```diff
@@ -81,18 +83,28 @@ func (node *Node) proposeNewBlock() (*types.Block, error) {
+	}
+
+	// TODO: integrate staking transaction into tx pool
+	pendingStakingTransactions := types2.StakingTransactions{}
```
do we need staking tx here?

LeoHChen commented on node/worker/worker.go:
```diff
@@ -50,99 +53,98 @@ type Worker struct {
+// CommitTransactions commits transactions for new block.
+func (w *Worker) CommitTransactions(
+	pendingNormal map[common.Address]types.Transactions,
+	pendingStaking staking.StakingTransactions, coinbase common.Address,
```
pendingStaking is not needed, right?

LeoHChen commented:
please also provide a test log on this change.

LeoHChen commented:
Thanks for all the clarification.  Please add some of the key clarification into the code and I am okay to approve it. Also, the gas fee changes have to be added to master branch as well.  Or maybe it is already changed in master branch. If not, we need to open an issue to keep track of it. Thanks.

rlan35 commented:
> Thanks for all the clarification. Please add some of the key clarification into the code and I am okay to approve it. Also, the gas fee changes have to be added to master branch as well. Or maybe it is already changed in master branch. If not, we need to open an issue to keep track of it. Thanks.

All the change in this PR are cherry-picked from master, so they are already done in master. Added comment. Also removed the staking txns arg in CommitTransactions method to be safe.

rlan35 commented:
> LGTM in general. I assume the txpool functions were tested in master/s3 branch.

Yea, basic txpool logic was tested in master when it was implemented. It's good we can test it more extensively for token swap now.
